### PR TITLE
[core] Cache record in append-only write using `ExternalBuffer`, avoid out-of-memory error while writing lots of partitions

### PR DESCRIPTION
**Pull request by leaves12138**
### Problem
  While writing to append-only table, we have no cache in append-only writer. If the records between two checkpoint interval too many, we will get a out-of-memory exception. In conclusion, it is not a good idea to cache all the records between two checkpoint in memory. So here, introduce ExternalBuffer to cache the record, and flush it while memory pool is exhausted.

### Influence 
1、 The write speed will be slowed down.  We have to serialize all the record to segment, and deserialize it while checkpoint comming.
2、The  checkpoint complete will be delayed.  If we spill record on disk, we have to read them and write it to a file.

### Discuss
Now `ExternalBuffer` default enable spill and have no parameter to disable that. Whether we need an option to disable spill function, and to be compatible with the table-config "write-buffer-spillable".  Or just enable spill on `ExternalBuffer` by default.

Please give me some advice about the discuss ...

### Tests
Added tests for `SpillableBuffer` and `AppendOnlyWrite`

### CODE REVIEW

**Comment by leaves12138** [comment]
Now I add the spillable config to SpillableBuffer (which is renamed from ExternalBuffer). And if spillable is false, append-only writer will flush the files to disk, which may cause small too many small files.

**Comment by leaves12138** [comment]
Please review my code and give me some options.

**Comment by wxplovecc** [comment]
Append only table write to files direct in my understanding, the OOM maybe caused by too many writers?

**Comment by JingsongLi** [comment]
I think we should keep directly writing for append table.

**Comment by leaves12138** [comment]
> Append only table write to files direct in my understanding, the OOM maybe caused by too many writers?

Records may cache in OrcWriter/ParquetWriter, and too many writers cache too many records cause the OOM

**Comment by leaves12138** [comment]
> I think we should keep directly writing for append table by default. For example, 256MB in memory buffer may can not write 128MB orc file, but previous behavior can.

reasonable

File: paimon-common/src/main/java/org/apache/paimon/CoreOptions.java [hunk 1]
```diff
@@ -277,6 +277,13 @@ public class CoreOptions implements Serializable {
                     .withDescription(
                             "Whether the write buffer can be spillable. Enabled by default when using object storage.");
 
+    public static final ConfigOption<Boolean> USE_WRITE_BUFFER =
+            key("use-write-buffer")
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/CoreOptions.java:281]: `write-buffer-for-append`?

Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/CoreOptions.java:281]: Also document this in append-only documentation page.

File: paimon-core/src/main/java/org/apache/paimon/disk/SpillableBuffer.java [hunk 2]
```diff
@@ -53,14 +53,18 @@ public class ExternalBuffer {
...
+    private final boolean spillable;
 
     private final List<ChannelWithMeta> spilledChannelIDs;
     private int numRows;
 
     private boolean addCompleted;
 
-    public ExternalBuffer(
-            IOManager ioManager, MemorySegmentPool pool, AbstractRowDataSerializer<?> serializer) {
+    public SpillableBuffer(
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/disk/SpillableBuffer.java:63]: Don't change this name... The renaming is not so solid.

File: paimon-core/src/main/java/org/apache/paimon/disk/SpillableBuffer.java [hunk 3]
```diff
@@ -53,14 +53,18 @@ public class ExternalBuffer {
...
+            IOManager ioManager,
+            MemorySegmentPool pool,
+            AbstractRowDataSerializer<?> serializer,
+            boolean spillable) {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/disk/SpillableBuffer.java:67]: You should first introduce a new interface, and ExternalBuffer and InMemoryBuffer implements it.

File: paimon-core/src/main/java/org/apache/paimon/sort/BinaryInMemorySortBuffer.java [hunk 4]
```diff
@@ -182,6 +182,7 @@ public boolean write(InternalRow record) throws IOException {
         long currOffset = currentDataBufferOffset + skip;
 
         writeIndexAndNormalizedKey(record, currOffset);
+        sortIndexBytes += indexEntrySize;
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/sort/BinaryInMemorySortBuffer.java:185]: This can be a separate PR.

File: paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyWriter.java [hunk 5]
```diff
@@ -63,12 +69,17 @@ public class AppendOnlyWriter implements RecordWriter<InternalRow> {
     private final List<DataFileMeta> compactAfter;
     private final LongCounter seqNumCounter;
     private final String fileCompression;
+    private final boolean useWriteBuffer;
+    private final boolean spillable;
     private final FieldStatsCollector.Factory[] statsCollectors;
+    private final IOManager ioManager;
 
+    private SpillableBuffer writeBuffer;
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyWriter.java:77]: You can create a small inner `interface`. And with two small inner class implement it.

File: paimon-common/src/main/java/org/apache/paimon/CoreOptions.java (hunk 1 above)
Comment (leaves12138) [reply paimon-common/src/main/java/org/apache/paimon/CoreOptions.java:281]: done

File: paimon-core/src/main/java/org/apache/paimon/disk/SpillableBuffer.java (hunk 3 above)
Comment (leaves12138) [reply paimon-core/src/main/java/org/apache/paimon/disk/SpillableBuffer.java:67]: done

File: paimon-core/src/main/java/org/apache/paimon/sort/BinaryInMemorySortBuffer.java (hunk 4 above)
Comment (leaves12138) [reply paimon-core/src/main/java/org/apache/paimon/sort/BinaryInMemorySortBuffer.java:185]: reverted

File: paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyWriter.java (hunk 5 above)
Comment (leaves12138) [reply paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyWriter.java:77]: done

File: paimon-common/src/main/java/org/apache/paimon/CoreOptions.java (hunk 1 above)
Comment (leaves12138) [reply paimon-common/src/main/java/org/apache/paimon/CoreOptions.java:281]: done

File: paimon-core/src/main/java/org/apache/paimon/disk/SpillableBuffer.java (hunk 2 above)
Comment (leaves12138) [reply paimon-core/src/main/java/org/apache/paimon/disk/SpillableBuffer.java:63]: reverted

**Review by JingsongLi — APPROVED** [review]
+1

**Review by JingsongLi — APPROVED** [review]
+1
